src/rfc3161_client/verify.py

```diff
@@ -12,170 +13,267 @@
-
-    :param sig: Bytes of a PKCS7 object. This must be in DER format and will be unserialized.
-    :param certificates: A list of trusted certificates to verify the response against.
-    :raise: ValueError if the signature verification fails.
-    """
-    return _rust_verify.pkcs7_verify(sig, list(certificates))
-
-
-def _verify_leaf_certs(tsp_response: TimeStampResponse, opts: VerifyOpts) -> bool:
-    if opts.tsa_certificate is None and len(tsp_response.signed_data.certificates) == 0:
-        msg = "Certificates neither found in the answer or in the opts."
-        raise VerificationError(msg)
-
-    if len(tsp_response.signed_data.certificates) > 0:
-        leaf_certificate_bytes = next(iter(tsp_response.signed_data.certificates))
-        leaf_certificate = cryptography.x509.load_der_x509_certificate(leaf_certificate_bytes)
-
-        if opts.tsa_certificate is not None and leaf_certificate != opts.tsa_certificate:
-            msg = "Embedded certificate does not match the one in the opts."
+class VerifierBuilder:
+    def __init__(
+        self,
+        policy_id: cryptography.x509.ObjectIdentifier | None = None,
+        tsa_certificate: cryptography.x509.Certificate | None = None,
+        intermediates: list[cryptography.x509.Certificate] | None = None,
+        roots: list[cryptography.x509.Certificate] | None = None,
+        nonce: int | None = None,
+        common_name: str | None = None,
+    ):
+        """Init method."""
+        self._policy_id: cryptography.x509.ObjectIdentifier | None = policy_id
+        self._tsa_certificate: cryptography.x509.Certificate | None = tsa_certificate
+        self._intermediates: list[cryptography.x509.Certificate] = intermediates or []
+        self._roots: list[cryptography.x509.Certificate] = roots or []
+        self._nonce: int | None = nonce
+        self._common_name: str | None = common_name
+
+    def policy_id(self, policy_oid: cryptography.x509.ObjectIdentifier) -> VerifierBuilder:
+        """Set the policy ID."""
+        if self._policy_id is not None:
+            msg = "The policy id can be set only once"
+            raise ValueError(msg)
+        builder = copy(self)
+        builder._policy_id = policy_oid
+        return builder
+
+    def tsa_certificate(self, certificate: cryptography.x509.Certificate) -> VerifierBuilder:
+        """Set the TSA certificate."""
+        if self._tsa_certificate is not None:
+            msg = "The TSA certificate can be set only once"
+            raise ValueError(msg)
+        builder = copy(self)
+        builder._tsa_certificate = certificate
+        return builder
+
+    def add_intermediate_certificate(
+        self, certificate: cryptography.x509.Certificate
+    ) -> VerifierBuilder:
+        """Add an intermediate certificate."""
+        intermediates = self._intermediates
+        if certificate in intermediates:
+            msg = "The certificate is already present"
```

Comment:
```suggestion
            msg = "intermediate certificate is already present"
```

Reply:
NB: Technically supplying the intermediate more than once isn't really an issue, since the chain building process will ignore it. But I suppose it doesn't really hurt to restrict it either.